zcash_client_sqlite/src/wallet/transparent.rs

```diff
@@ -32,3 +50,755 @@ pub(crate) fn detect_spending_accounts<'a>(
+    let chain_tip_height = chain_tip_height(conn)?.ok_or(SqliteClientError::ChainHeightUnknown)?;
+
+    let mut stmt_address_balances = conn.prepare(
+        "SELECT u.address, SUM(u.value_zat)
+         FROM transparent_received_outputs u
+         JOIN transactions t
+         ON t.id_tx = u.transaction_id
+         WHERE u.account_id = :account_id
+         -- the transaction that created the output is mined or is definitely unexpired
+         AND (
+            t.mined_height <= :summary_height -- tx is mined
+            OR ( -- or the caller has requested to include zero-conf funds that are not expired
+                :summary_height > :chain_tip_height
+                AND (
+                    t.expiry_height = 0 -- tx will not expire
+                    OR t.expiry_height >= :summary_height
+                )
+            )
+         )
+         -- and the output is unspent
+         AND u.id NOT IN (
+            SELECT txo_spends.transparent_received_output_id
+            FROM transparent_received_output_spends txo_spends
+            JOIN transactions tx ON tx.id_tx = txo_spends.transaction_id
+            WHERE tx.mined_height IS NOT NULL -- the spending tx is mined
+            OR tx.expiry_height = 0 -- the spending tx will not expire
+            OR tx.expiry_height >= :spend_expiry_height -- the spending tx is unexpired
+         )
+         GROUP BY u.address",
+    )?;
+
+    let mut res = HashMap::new();
+    let mut rows = stmt_address_balances.query(named_params![
+        ":account_id": account.0,
+        ":summary_height": u32::from(summary_height),
+        ":chain_tip_height": u32::from(chain_tip_height),
+        ":spend_expiry_height": u32::from(std::cmp::min(summary_height, chain_tip_height + 1)),
+    ])?;
+    while let Some(row) = rows.next()? {
+        let taddr_str: String = row.get(0)?;
+        let taddr = TransparentAddress::decode(params, &taddr_str)?;
+        let value = NonNegativeAmount::from_nonnegative_i64(row.get(1)?)?;
+
+        res.insert(taddr, value);
+    }
+
+    Ok(res)
+}
+
+#[tracing::instrument(skip(conn, account_balances))]
+pub(crate) fn add_transparent_account_balances(
+    conn: &rusqlite::Connection,
+    mempool_height: BlockHeight,
+    account_balances: &mut HashMap<AccountId, AccountBalance>,
+) -> Result<(), SqliteClientError> {
+    let transparent_trace = tracing::info_span!("stmt_transparent_balances").entered();
+    let mut stmt_account_balances = conn.prepare(
+        "SELECT u.account_id, SUM(u.value_zat)
+         FROM transparent_received_outputs u
+         JOIN transactions t
+         ON t.id_tx = u.transaction_id
+         -- the transaction that created the output is mined or is definitely unexpired
+         WHERE (
+            t.mined_height < :mempool_height -- tx is mined
+            OR t.expiry_height = 0 -- tx will not expire
+            OR t.expiry_height >= :mempool_height
+         )
```

Comment:
Observation: The other  minconf-0 instances of this pattern use `t.mined_height <= SOME_HEIGHT OR t.expiry_height >= SOME_HEIGHT`. I think it's okay to use `< :mempool_height` here however, solely because we never actually encounter the off-by-one here (because we encode mempool heights in the database as `NULL`).